cmd/kubeturbo/app/kubeturbo_builder.go

```diff
@@ -1009,3 +1011,47 @@ func reviewSCCAccess(namespace string, kubeClient kubernetes.Interface) bool {
 
 	return true
 }
+
+func WatchConfigMap() {
+	//Check if the file /etc/kubeturbo/turbo-autoreload.config exists
+	autoReloadConfigFilePath := "/etc/kubeturbo"
+	autoReloadConfigFileName := "turbo-autoreload.config"
+
+	viper.AddConfigPath(autoReloadConfigFilePath)
+	viper.SetConfigType("json")
+	viper.SetConfigName(autoReloadConfigFileName)
+	for {
+		verr := viper.ReadInConfig()
+		if verr == nil {
+			break
+		} else {
+			glog.V(4).Infof("Can't read the autoreload config file %s/%s due to the error: %v, will retry in 3 seconds", autoReloadConfigFilePath, autoReloadConfigFileName, verr)
+			time.Sleep(3 * time.Second)
+		}
+	}
+
+	glog.V(1).Infof("Start watching the autoreload config file %s/%s", autoReloadConfigFilePath, autoReloadConfigFileName)
+	updateConfig := func() {
+		newLoggingLevel := viper.GetString("logging.level")
+		currentLoggingLevel := pflag.Lookup("v").Value.String()
+		if newLoggingLevel != currentLoggingLevel {
+			if newLogVInt, err := strconv.Atoi(newLoggingLevel); err != nil || newLogVInt < 0 {
+				glog.Errorf("Invalid log verbosity %v in the autoreload config file", newLoggingLevel)
+			} else {
+				err := pflag.Lookup("v").Value.Set(newLoggingLevel)
+				if err != nil {
+					glog.Errorf("Can't apply the new logging level setting due to the error:%v", err)
+				} else {
+					glog.V(1).Infof("Logging level is changed from %v to %v", currentLoggingLevel, newLoggingLevel)
+				}
+
+			}
+		}
+	}
+	updateConfig() //update the logging level during startup
+	viper.OnConfigChange(func(in fsnotify.Event) {
+		updateConfig()
+	})
+
+	viper.WatchConfig()
```

Comment:
I believe this is expected to block. Is it possible this can return unexpectedly? if its possible we need a for loop here to restart the watch.

Reply:
> 

This is the implementation of the `watchConfig` function.  It doesn't block, it starts another Go routine and keeps watching the file event.  It doesn't return an error,  all of the errors are handled internally.  
```
func (v *Viper) WatchConfig() {
	initWG := sync.WaitGroup{}
	initWG.Add(1)
	go func() {
		watcher, err := newWatcher()
		if err != nil {
			log.Fatal(err)
		}
		defer watcher.Close()
		// we have to watch the entire directory to pick up renames/atomic saves in a cross-platform way
		filename, err := v.getConfigFile()
		if err != nil {
			log.Printf("error: %v\n", err)
			initWG.Done()
			return
		}

		configFile := filepath.Clean(filename)
		configDir, _ := filepath.Split(configFile)
		realConfigFile, _ := filepath.EvalSymlinks(filename)

		eventsWG := sync.WaitGroup{}
		eventsWG.Add(1)
		go func() {
			for {
				select {
				case event, ok := <-watcher.Events:
					if !ok { // 'Events' channel is closed
						eventsWG.Done()
						return
					}
					currentConfigFile, _ := filepath.EvalSymlinks(filename)
					// we only care about the config file with the following cases:
					// 1 - if the config file was modified or created
					// 2 - if the real path to the config file changed (eg: k8s ConfigMap replacement)
					const writeOrCreateMask = fsnotify.Write | fsnotify.Create
					if (filepath.Clean(event.Name) == configFile &&
						event.Op&writeOrCreateMask != 0) ||
						(currentConfigFile != "" && currentConfigFile != realConfigFile) {
						realConfigFile = currentConfigFile
						err := v.ReadInConfig()
						if err != nil {
							log.Printf("error reading config file: %v\n", err)
						}
						if v.onConfigChange != nil {
							v.onConfigChange(event)
						}
					} else if filepath.Clean(event.Name) == configFile &&
						event.Op&fsnotify.Remove&fsnotify.Remove != 0 {
						eventsWG.Done()
						return
					}

				case err, ok := <-watcher.Errors:
					if ok { // 'Errors' channel is not closed
						log.Printf("watcher error: %v\n", err)
					}
					eventsWG.Done()
					return
				}
			}
		}()
		watcher.Add(configDir)
		initWG.Done()   // done initializing the watch in this go routine, so the parent routine can move on...
		eventsWG.Wait() // now, wait for event loop to end in this go-routine...
	}()
	initWG.Wait() // make sure that the go routine above fully ended before returning
}
```